adding confirm_req to_string()

not-Ian opened the pull request:
re: #3960

not-Ian committed:
adding to_string to confirm_req

dsiganos commented on nano/node/messages.cpp:
```diff
@@ -940,6 +940,21 @@ std::size_t nano::confirm_req::size (nano::block_type type_a, std::size_t count)
+
+	stream << header.to_string ();
+
+	for (auto roots_hash = roots_hashes.begin (), end = roots_hashes.end (); roots_hash != end; ++roots_hash)
```
This line is too complicated. Please use range loop to simplify this code.
https://en.cppreference.com/w/cpp/language/range-for

not-Ian replied:
this should be resolved after the recent push

dsiganos commented on nano/core_test/message.cpp:
```diff
@@ -243,6 +243,29 @@ TEST (message, confirm_req_hash_batch_serialization)
 	ASSERT_EQ (header.count_get (), req.roots_hashes.size ());
 }
 
+TEST (message, confirm_req_to_string)
```
I asked for unit tests and thank you for doing them but let's make them minimal.
It is not worth having and maintaining large amount of code testing to_string functions.
Please re-write this to be minimal.

dsiganos commented on nano/core_test/message.cpp:
```diff
@@ -243,6 +243,29 @@ TEST (message, confirm_req_hash_batch_serialization)
+	nano::send_block block = nano::send_block ();
+	std::shared_ptr block_ptr = std::make_shared<nano::send_block> (block);
+
+	nano::work_thresholds work_threshold = nano::work_thresholds (0, 0, 0);
```
There is a much more concise way to do this, which we used in the test case:
'message.keepalive_to_string'. Please use that approach.

dsiganos commented on nano/core_test/message.cpp:
```diff
@@ -243,6 +243,29 @@ TEST (message, confirm_req_hash_batch_serialization)
+
+	std::string expected_string = "NetID: 5241(dev), VerMaxUsingMin: 19/19/18, MsgType: 4(confirm_req), Extensions: 0200";
+
+	for (auto roots_hash = 0, roots_hash_end = 3; roots_hash != roots_hash_end; ++roots_hash)
```
There is no need for a for loop here. This is test code, expand the loop and do it manually.

dsiganos commented on nano/node/messages.cpp:
```diff
@@ -940,6 +940,21 @@ std::size_t nano::confirm_req::size (nano::block_type type_a, std::size_t count)
 	return result;
 }
 
+std::string nano::confirm_req::to_string () const
```
confirm_req has 2 modes of operation:
* list of hash pair
* block to confirm
You are only handling the list of pairs mode.
You need to handle both modes.

The block_type in the header determines which mode it is in.

not-Ian committed:
simplified the test, included confirm_req with a block instead of roots_hashes use case

not-Ian committed:
formatting

dsiganos committed:
Using a string instead of a stringstream and using colon as separator

1. There seems to be no reason to use stringstream over just a plain string.
2. I am using a colon to separate the root and hash because that seems to be
a format we use a lot.

dsiganos committed:
Remove the unit tests

In this case, it is not worth having so much non trivial test code just
to test a to_string function.
